snap/hooks/configure

```diff
@@ -1,11 +1,80 @@
 #!/usr/bin/env python3
 
+"""Clean all related obsoleted and outdated LXD base images and instances
+
+This hook will ensure only exists one latest base instance every 
+(craft base metadata version, image provider, image os, image version)
+tuple. All other bases and below minimal base versions image/instance
+will be deleted.
+
+Example:
+base-instance-charmcraft-buildd-base-v1.0-craft-com.ubuntu.cloud-buildd-core22
+                                       |          |                 |      |
+craft base metadata version: 1.0 -------          |                 |      |
+image provider: craft-com.ubuntu.cloud ------------                 |      |
+image os: buildd ----------------------------------------------------      |
+image version: core22 ------------------------------------------------------
+"""
+
+import json
+import re
+import subprocess
 import sys
 
+from datetime import datetime
+
 from charmcraft import snap
 
+PROJECT_NAME = "charmcraft"
+MINIMAL_BASE_VERISON = "1.0"
+
+# We need craft metadata version, image provider, image os, image version from regex
+# base-instance-charmcraft-buildd-base-v1.0-craft-com.ubuntu.cloud-buildd-core22
+SUPPORTED_VERSIONS = [
+    r"base-instance-" + re.escape(PROJECT_NAME) + r"-\w+-base-v(\d+\.\d+)-([\w\.\-]+)-(\w+)-(\w+)"
+]
+
+# snapshot-craft-com.ubuntu.cloud-buildd-core22-charmcraft-buildd-base-v0.0
+# base-instance-charmcraft-buildd-base-v10-689fe40991089812a57b
+OBSOLETE_VERSIONS = [
+    r"snapshot-craft-[\w\.]+-\w+-\w+-" + re.escape(PROJECT_NAME) + r"-\w+-base-v[\d\.]+",
+    r"base-instance-" + re.escape(PROJECT_NAME) + r"-(\w+)-base-v([\d\.]+)-\w+",
+]
+
+
+def _delete_lxd_image(image: dict) -> None:
+    print(
+        f"Removing old image {image['aliases'][0]['name']} ({image['fingerprint']}) "
+        f"in LXD {PROJECT_NAME} project..."
+    )
+    try:
+        subprocess.check_output(
+            [
+                "lxc",
+                "--project",
+                PROJECT_NAME,
+                "image",
+                "delete",
+                image["fingerprint"],
+            ]
+        )
+    except subprocess.CalledProcessError:
+        print(f"Remove LXD image {image['fingerprint']} failed.", file=sys.stderr)
+
+
+def _delete_lxd_instance(instance: dict) -> None:
+    print(f"Removing old instance {instance['name']} in LXD {PROJECT_NAME} project...")
+    try:
+        subprocess.check_output(
+            ["lxc", "--project", PROJECT_NAME, "delete", "--force", instance["name"]]
+        )
+    except subprocess.CalledProcessError:
+        print(f"Remove LXD instance {instance['name']} failed.", file=sys.stderr)
+
 
 def configure_hook_main():
+    image_slots = {}
```

Comment:
Can you add a comment about what this variable is? (It's not used until later in the file, so it would be nice to add a small hint here)